docs/network-drivers.md

```diff
@@ -1,47 +1,127 @@
+for developers writing tools which interact with the API as well.
+
+# Overview
+
+There are two classes of network-related drivers: network allocators and
+switches. Switches are straightforward to understand; they provide
+support for a particular hardware switch. Network allocators are a
+bit more abstract.
+
+HaaS networks may be backed by various underlying isolation technologies
+(currently the only network allocator shipped with HaaS uses 802.1q
+VLANs, but there may be more, e.g. VXLANs in the future). A network
+allocator manages the details of mapping networks created by HaaS to
+these underlying technologies. For example, the ``vlan_pool`` allocator
+maps each network to a unique VLAN id.
+
+Drivers are implemented as extensions, and must be added to the
+``[extensions]`` section of ``haas.cfg``. You must supply *exactly* one
+network allocator driver, and one or more switches. For example::
+
+    ...
+    [extensions]
+    haas.ext.network_allocators.vlan_pool =
+    haas.ext.switches.dell =
+    haas.ext.switches.nexus =
+
+Some drivers may also need driver specific options, which go in a
+section with the same name as the extension, e.g.::
+
+    [haas.ext.network_allocators.vlan_pool]
+    vlans = 300-500, 700-750
+
+# Network allocator drivers
+
+The only network allocator shipped with HaaS that is of interest to
+users (there are others useful for development purposes) is the VLAN
+allocator. The name of the extension is
+``haas.ext.network_allocators.vlan_pool``, and it requires a single
+extension-specific config option, `vlans`, which is a comma separated
+list of VLAN ids and/or ranges of VLAN ids. Networks created within HaaS
+will use VLANs specified in the configuration file. An example::
+
+    ...
+
+    [extensions]
+    haas.ext.network_allocators.vlan_pool =
+    ...
+
+    [haas.ext.network_allocators.vlan_pool]
+    vlans = 300, 500-700, 800-950
+    ...
+
+# Switch drivers
+
+At present, all switch drivers shipped with HaaS require that the VLAN
+pool allocator is in use. There are two switch drivers shipped with
+HaaS:
+
+* ``haas.ext.switches.dell``, which provides a driver for the Dell
+  Powerconnect 5500 series switches.
+* ``haas.ext.switches.nexus``, which provides a driver for some Cisco
+  Nexus switches. Only the 3500 and 5500 have been tested, though it is
+  possible that other models will work as well.
+
+Neither driver requires any extension-specific config options. Per the
+information in `rest_api.md`, the details of certain API calls are
+driver-dependant, below are the details for each of these switches.
+
+## Powerconnect driver
+
+### switch_register
+
+To register a Dell Powerconnect switch, the ``"type"`` field of the
+request body must have a value of::
+
+    http://schema.massopencloud.org/haas/switches/powerconnect55xx
+
+In addition, it requires three extra fields: ``"username"``,
+``"hostname"``, and ``"password"``, which provide the necessary
```

Comment:
They could also do anything they want to the nodes' IPMI controllers.

I'm disinclined to spent too much time on this; separate api/network daemon servers are still something we can do in theory, but have never tried nor felt the need to devote the resources to. I think any of these solutions is going to involve a substantial amount of work, and it will only have any impact at all in a use case that we haven't actually had need for.